lib/rules/no-inner-declarations.js

```diff
@@ -97,8 +113,24 @@ module.exports = {
 
         return {
 
-            FunctionDeclaration: check,
+            FunctionDeclaration(node) {
+                const isStrictFunction = sourceCode.getScope(node).isStrict;
+
+                if (isStrictFunction && legacy === false) {
+                    return;
+                }
+
+                if (isStrictFunction && ecmaVersion !== 5) {
+                    return;
+                }
+
+                check(node);
+            },
             VariableDeclaration(node) {
+                if (isStrict) {
```

Comment:
We should check strict mode here again, because a variable can be inside of a function that is in strict mode even if the entire script is not.
```suggestion
                if (sourceCode.getScope(node).isStrict) {
```